packages/stg-evm-v2/devtools/src/token-messaging/state.ts

```diff
@@ -0,0 +1,147 @@
+        planner: (await sdk.getPlanner()) ?? '0x',
+        maxAssetId: (await sdk.getMaxAssetId()) ?? 0,
+        assets: onChainAssets,
+    }
+    return { point, config }
+}
+
+async function getEdgeFromChain(
+    vector: OmniVector,
+    existingConfig: TokenMessagingEdgeConfig
+): Promise<OmniEdge<TokenMessagingEdgeConfig>> {
+    const logger = createModuleLogger(`State Loader for TokenMessaging`)
+    const createSdk = createTokenMessagingFactory(createConnectedContractFactory())
+    const createOappSdk = createOAppFactory(createConnectedContractFactory())
+    const sdk = await createSdk(vector.from)
+    const oappSdk = await createOappSdk(vector.from)
+    const endpointSdk = await oappSdk.getEndpointSDK()
+
+    logger.debug(`Querying edges on ${getNetworkNameForEid(vector.from.eid)}...`)
+
+    const [sendLibrary, recvLibrary] = await Promise.all([
+        endpointSdk.getSendLibrary(vector.from.address, vector.to.eid),
+        endpointSdk.getReceiveLibrary(vector.from.address, vector.to.eid),
+    ])
+
+    const [executorConfig, sendUlnConfig] = sendLibrary
+        ? await Promise.all([
+              endpointSdk.getExecutorConfig(vector.from.address, sendLibrary ?? '0x', vector.to.eid),
+              endpointSdk.getUlnConfig(vector.from.address, sendLibrary ?? '0x', vector.to.eid),
+          ])
+        : [undefined, undefined]
+
+    const recvUlnConfig = recvLibrary
+        ? await endpointSdk.getUlnConfig(vector.from.address, recvLibrary[0] ?? '0x', vector.to.eid)
+        : { requiredDVNs: undefined }
+
+    const [maxPassengers, gasLimit, nativeDropAmount] = await Promise.all([
+        sdk.getMaxPassengers(vector.to.eid),
+        sdk.getGasLimit(vector.to.eid),
+        sdk.getNativeDropAmount(vector.to.eid),
+    ])
+
+    const config: TokenMessagingEdgeConfig = {
+        // Fares not part of the config, they are set by the planner
+        enforcedOptions: [],
+        maxPassengers: maxPassengers ?? 0,
+        gasLimit: gasLimit ?? { gasLimit: 0n, nativeDropGasLimit: 0n },
+        nativeDropAmount: nativeDropAmount ?? 0n,
+        sendConfig: {
+            executorConfig,
+            ulnConfig: sendUlnConfig, // confirmations is not in the config
+        },
+        receiveConfig: { ulnConfig: { requiredDVNs: recvUlnConfig?.requiredDVNs ?? [] } },
+    }
+
+    if (!existingConfig.enforcedOptions) {
+        logger.debug(`No enforced options for ${getNetworkNameForEid(vector.to.eid)}`)
+        return { vector, config }
+    }
+
+    const enforcedOptions: OAppEnforcedOption[] = []
+    for (const enforcedOptionsConfig of existingConfig.enforcedOptions) {
+        const optionsHex = await oappSdk.getEnforcedOptions(vector.to.eid, enforcedOptionsConfig.msgType)
+        if (isZero(optionsHex)) continue
+
+        const options = Options.fromOptions(optionsHex)
+        enforcedOptions.push(...decodeEnforcedOptions(options, enforcedOptionsConfig.msgType))
+    }
+    config.enforcedOptions = enforcedOptions
```

Comment:
This you could make more functional & parallelized by doing smth like:

```typescript
const enforcedOptions = await Promise.all(existingConfig.enforcedOptions.map(async () => {
    const optionsHex = await oappSdk.getEnforcedOptions(vector.to.eid, enforcedOptionsConfig.msgType)
    if (isZero(optionsHex)) return []

    const options = Options.fromOptions(optionsHex)
    return decodeEnforcedOptions(options, enforcedOptionsConfig.msgType)
})

config.enforcedOptions = enforcedOptions.flat()
```

Reply:
Thank you!